devmand/gateway/src/devmand/channels/cli/KeepaliveCli.cpp

```diff
@@ -0,0 +1,136 @@
+    shared_ptr<Cli> _cli,
+    shared_ptr<Executor> _parentExecutor,
+    shared_ptr<ThreadWheelTimekeeper> _timekeeper,
+    chrono::milliseconds _heartbeatInterval,
+    Command&& _keepAliveCommand,
+    chrono::milliseconds _backoffAfterKeepaliveTimeout)
+    : id(_id),
+      cli(_cli),
+      timekeeper(_timekeeper),
+      parentExecutor(_parentExecutor),
+      serialExecutorKeepAlive(SerialExecutor::create(
+          Executor::getKeepAliveToken(_parentExecutor.get()))),
+      keepAliveCommand(_keepAliveCommand),
+      heartbeatInterval(_heartbeatInterval),
+      backoffAfterKeepaliveTimeout(_backoffAfterKeepaliveTimeout) {
+  assert(_keepAliveCommand.skipCache());
+  shutdown = false;
+
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "initialized";
+}
+
+KeepaliveCli::~KeepaliveCli() {
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "Destructor started";
+  shutdown = true;
+  serialExecutorKeepAlive = nullptr;
+  parentExecutor = nullptr;
+  cli = nullptr;
+  timekeeper = nullptr;
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "Destructor done";
+}
+
+void KeepaliveCli::sendKeepAliveCommand() {
+  if (shutdown)
+    return;
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "sendKeepAliveCommand";
+  Future<string> result =
+      via(serialExecutorKeepAlive).thenValue([dis = shared_from_this()](auto) {
+        if (dis->shutdown)
+          throw runtime_error("KACli: Shutting down");
+        MLOG(MDEBUG) << "[" << dis->id << "] "
+                     << "sendKeepAliveCommand executing keepalive command";
+        return dis->executeAndRead(dis->keepAliveCommand);
+      });
+
+  scheduleNextPing(move(result));
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "sendKeepAliveCommand() done";
+}
+
+void KeepaliveCli::scheduleNextPing(Future<string> keepAliveCmdFuture) {
+  if (shutdown)
+    return;
+  MLOG(MDEBUG) << "[" << id << "] "
+               << "scheduleNextPing";
+  move(keepAliveCmdFuture)
+      .via(serialExecutorKeepAlive)
+      .thenValue([dis = shared_from_this()](auto) -> SemiFuture<Unit> {
+        MLOG(MDEBUG) << "[" << dis->id << "] "
+                     << "Creating sleep future";
+        return futures::sleep(dis->heartbeatInterval, dis->timekeeper.get());
+      })
+      .thenValue([dis = shared_from_this()](auto) -> Unit {
+        MLOG(MDEBUG) << "[" << dis->id << "] "
+                     << "Woke up after sleep";
+        dis->sendKeepAliveCommand();
+        return Unit{};
+      })
+      .thenError(
+          folly::tag_t<std::exception>{},
+          [dis = shared_from_this()](std::exception const& e) -> Future<Unit> {
+            MLOG(MINFO)
+                << "[" << dis->id << "] "
+                << "Got error running keepalive, backing off "
+                << e.what(); // FIXME: real exception is not propagated here
+            if (dis->shutdown)
+              return Unit{}; // sleep might hang
+            std::this_thread::sleep_for(
```

Comment:
Should we back off on any exception ? 

There can be: Execution timeout, Cli stack below closing, Session closed ... in the future even "Error response from device detected".

So in which case should we be applying the backoff ? Maybe we should be expecting only a special ex type for backoff.

Reply:
Agreed, either we should define our exceptions or use some meaningful result types. We'll discuss this elsewhere.